### DISCUSSION TITLE
Mismatch of calculated profit to score

### DISCUSSION THREAD

**Discussion by rnz1234** [yasserfarouk/scml#61] [one-shot-qa] [answered]
In the subject of utility/profit, in case I want to calculate the profit per negotiation myself:

1. Can I say that the profit obtained by the agent on a successful negotiation is:

if it is an L0 (negotiates on selling):
(contract.agreement["unit_price"]-float(self.awi.current_exogenous_input_price)/self.awi.current_exogenous_input_quantity-self.awi.profile.cost)*contract.agreement["quantity"]
 
if it is an L1 (negotiates on buying):(float(self.awi.current_exogenous_output_price)/self.awi.current_exogenous_output_quantity-contract.agreement["unit_price"]-self.awi.profile.cost)*contract.agreement["quantity"]#*self.awi.current_exogenous_output_quantity

2. Penalty:
After simulation step is finished, is the accumulated profit of the agent can be calculated by:

if I'm L0:
    too_much_penalty = self.awi.current_shortfall_penalty
else:
    too_much_penalty = self.awi.current_disposal_cost

if I'm L1:
    too_less_penalty = self.awi.current_disposal_cost
else:
    too_less_penalty = self.awi.current_shortfall_penalty        

if I bought too much:
    accum_profit += profit+self._needed()*too_much_penalty
else:
    accum_profit += profit-self._needed()*too_less_penalty

Currently I have a mismatch, I see that with the above profit calculation, the agent succeeds to "win" LearningAgent.
However the LearningAgent gets the higher score in the simulation. Can you advise what I'm doing wrong ?
is it better to use ufun to calculate the profit only on the contract signed in on_nego..success ?

Thanks

Thanks

**Comment by yasserfarouk** [comment, marked as answer]
> Can I say that the profit obtained by the agent on a successful negotiation is: ....

There are some issues with your approach:
1. The utility/profit is not even defined for a single contract. Consider the following case, you are an $L_0$ agent with 2 customers. You have exogenous supply of 10 items at 10 dollars each and your production cost, is 5. Now what is the utility of a contract to sell 6 items at a price of 20 to customer 1? We simply do not know because we must also know the other contract with customer 2 (if there is one) or the fact that *there are no other contracts* if that was the case.
2. Your calculation does not take into account the balance. May be you do not have enough money to buy everything or to produce everything you bought.
3. It does assume that you can convert all inputs to outputs. May be you can but some times you should not. The reason is that you may not have enough sale contracts. For example you may have 10 units of supply but sales for only 5. In such cases, you may just pay production cost for 5 items instead of paying for the 10 then paying again to dispose of the 5 units you created.


You can find one correct way to calculate the profit after a simulation in [from_offers() starting at line 201 in this file](https://github.com/yasserfarouk/scml/blob/master/src/scml/oneshot/ufun.py). As you can see, the calculation is involved to take account of all possibilities. This implementation matches the description of the ufun in the [Section 4 of the game description](https://github.com/yasserfarouk/scml/blob/master/src/scml/oneshot/ufun.py).

> However the LearningAgent gets the higher score in the simulation. Can you advise what I'm doing wrong ?

For the reasons mentioned above, the calculation you use is not very accurate. that is why it does not really match what the simulator is doing.

> is it better to use ufun to calculate the profit only on the contract signed in on_nego..success ?

To calculate the profit of any set of contracts, use ```self.ufun.from_contracts(contracts)```. This will always give you the correct answer assuming that the contracts you give to it are all executed and that no other contracts are executed in the same step. 

You can create **estimates** of the profit associated with  a single contract by predicting what will happen in the other negotiations, create contracts to represent your estimate, and pass all of the contracts (the one you are considering and the ones you estimated) together to `self.ufun.from_contracts`.
